xarray/core/indexing.py

```diff
@@ -432,16 +447,35 @@ def __init__(self, key):
         super().__init__(new_key)
 
 
-class ExplicitlyIndexed:
+T_ExplicitlyIndexed = TypeVar("T_ExplicitlyIndexed", bound="ExplicitlyIndexed")
+
+
+class ExplicitlyIndexed(metaclass=ABCMeta):
     """Mixin to mark support for Indexer subclasses in indexing."""
 
     __slots__ = ()
 
+    @abstractmethod
+    def __array__(self, dtype: DTypeLike = None) -> np.ndarray:
+        ...
+
+    @abstractmethod
+    def __getitem__(
+        self: T_ExplicitlyIndexed, key: Any
+    ) -> T_ExplicitlyIndexed | NumpyIndexingAdapter | np.ndarray | np.datetime64 | np.timedelta64:
+        ...
+        # TODO: fix the return type? Something more generic?
```

Comment:
We should really figure this out! I would love if we could constrain Xarray's interface more tightly here to simplify the programming model and reduce bugs. Type annotations are probably key!